// angular-reactive-form/esm2015/lib/directives/custom-checkbox/custom-checkbox.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, forwardRef, ElementRef, Renderer } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
export class CustomCheckboxComponent {
    /**
     * @param {?} element
     * @param {?} renderer
     */
    constructor(element, renderer) {
        this.element = element;
        this.renderer = renderer;
        this.onChang = (val) => { };
        this.onTouched = () => { };
        this.disabled = false;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    writeValue(val) {
        if (val) {
            this.value = val;
        }
    }
    /**
     * @return {?}
     */
    get value() {
        return this._value;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set value(val) {
        this._value = val;
        this.onChang(this._value);
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.onChang = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    setDisabledState(isDisabled) {
        this.disabled = isDisabled;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.value = false;
        console.log('value', this.value);
    }
    /**
     * @return {?}
     */
    onChangeFunction() {
        this.onChange();
    }
}
CustomCheckboxComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'custom-checkbox',
                template: "<input type=\"checkbox\" [(ngModel)]=\"value\" (change)=\"onChangeFunction()\">\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => CustomCheckboxComponent),
                        multi: true
                    }
                ],
                styles: [""]
            }] }
];
CustomCheckboxComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer }
];
CustomCheckboxComponent.propDecorators = {
    _value: [{ type: Input }],
    onChange: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    CustomCheckboxComponent.prototype._value;
    /** @type {?} */
    CustomCheckboxComponent.prototype.disabled;
    /** @type {?} */
    CustomCheckboxComponent.prototype.onChang;
    /** @type {?} */
    CustomCheckboxComponent.prototype.onTouched;
    /** @type {?} */
    CustomCheckboxComponent.prototype.onChange;
    /** @type {?} */
    CustomCheckboxComponent.prototype.element;
    /** @type {?} */
    CustomCheckboxComponent.prototype.renderer;
}
