// angular-reactive-form/esm2015/lib/models/UploadFile.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { InputBase } from './InputBase';
export class UploadFile extends InputBase {
    /**
     * @param {?=} options
     */
    constructor(options = {}) {
        super(options);
        this.controlType = 'uploadfile';
        this.accept = options['accept'] || '';
        this.allowMaxSize = options['allowMaxSize'] || 2;
        this.apiUpload = options['apiUpload'] || '';
        this.headers = options['headers'] || {};
        this.uploadPath = options['uploadPath'] || '';
        this.multiple = options['multiple'] || null;
        this.paramName = options['paramName'] || 'file';
        this.resultTransformer = options['resultTransformer'];
    }
}
if (false) {
    /** @type {?} */
    UploadFile.prototype.controlType;
    /** @type {?} */
    UploadFile.prototype.accept;
    /** @type {?} */
    UploadFile.prototype.allowMaxSize;
    /** @type {?} */
    UploadFile.prototype.apiUpload;
    /** @type {?} */
    UploadFile.prototype.headers;
    /** @type {?} */
    UploadFile.prototype.uploadPath;
    /** @type {?} */
    UploadFile.prototype.multiple;
    /** @type {?} */
    UploadFile.prototype.paramName;
    /** @type {?} */
    UploadFile.prototype.resultTransformer;
}
