// angular-reactive-form/esm2015/lib/directives/custom-selection/custom-selection.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, forwardRef, ElementRef, Renderer } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import * as _ from 'lodash';
export class CustomSelectionComponent {
    /**
     * @param {?} element
     * @param {?} renderer
     */
    constructor(element, renderer) {
        this.element = element;
        this.renderer = renderer;
        this.onChange = (val) => { };
        this.onTouched = () => { };
        this.disabled = false;
    }
    /**
     * @return {?}
     */
    triggerChanged() {
        /** @type {?} */
        let event = new CustomEvent('change', { bubbles: true });
        this.renderer.invokeElementMethod(this.element.nativeElement, 'dispatchEvent', [event]);
    }
    /**
     * @param {?} val
     * @return {?}
     */
    writeValue(val) {
        if (val) {
            this.value = val;
        }
    }
    /**
     * @return {?}
     */
    get value() {
        return this._value;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set value(val) {
        this.selected = _.find(this.options, item => item.value === val);
        this._value = val;
        this.onChange(this._value);
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.onChange = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    setDisabledState(isDisabled) {
        this.disabled = isDisabled;
    }
    /**
     * @param {?} opt
     * @return {?}
     */
    change(opt) {
        this.value = opt.value;
        this.triggerChanged();
    }
}
CustomSelectionComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'custom-selection',
                template: "<div class=\"dropdown custom-selection\">\n  <button class=\"btn btn-default dropdown-toggle\" type=\"button\" data-toggle=\"dropdown\" aria-haspopup=\"true\" aria-expanded=\"false\">\n    {{ selected === undefined ? '' : selected.label }}\n  </button>\n  <div class=\"dropdown-menu\">\n    <a class=\"dropdown-item\" *ngFor=\"let option of options\" (click)=\"change(option)\">{{ option.label }}</a>\n  </div>\n</div>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => CustomSelectionComponent),
                        multi: true
                    }
                ],
                styles: [".custom-selection .dropdown-menu .dropdown-item{cursor:pointer}"]
            }] }
];
CustomSelectionComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer }
];
CustomSelectionComponent.propDecorators = {
    _value: [{ type: Input }],
    options: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    CustomSelectionComponent.prototype._value;
    /** @type {?} */
    CustomSelectionComponent.prototype.options;
    /** @type {?} */
    CustomSelectionComponent.prototype.selected;
    /** @type {?} */
    CustomSelectionComponent.prototype.disabled;
    /** @type {?} */
    CustomSelectionComponent.prototype.onChange;
    /** @type {?} */
    CustomSelectionComponent.prototype.onTouched;
    /** @type {?} */
    CustomSelectionComponent.prototype.element;
    /** @type {?} */
    CustomSelectionComponent.prototype.renderer;
}
