// angular-reactive-form/esm2015/lib/directives/directives.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { RouterModule } from '@angular/router';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CustomSelectionComponent } from './custom-selection/custom-selection.component';
import { MaskInputComponent } from './mask-input/mask-input.component';
import { MutipleCheckboxComponent } from './mutiple-checkbox/mutiple-checkbox.component';
import { TextMaskModule } from 'angular2-text-mask';
import { CustomDropzoneComponent } from './dropzone/dropzone.component';
import { DropzoneModule } from 'ngx-dropzone-wrapper';
import { DROPZONE_CONFIG } from 'ngx-dropzone-wrapper';
import { CalendarModule } from 'primeng/calendar';
import { CustomDateTimePickerComponent } from './custom-date-time-picker/custom-date-time-picker.component';
import { PhoneInputComponent } from './phone-input/phone-input.component';
import { CustomSelect2Component } from './custom-select2/custom-select2.component';
import { CustomCheckboxComponent } from './custom-checkbox/custom-checkbox.component';
import { UploadImagesComponent } from './upload-images/upload-images.component';
/** @type {?} */
const DEFAULT_DROPZONE_CONFIG = {};
const ɵ0 = DEFAULT_DROPZONE_CONFIG;
export class DirectivesModule {
}
DirectivesModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    ReactiveFormsModule,
                    RouterModule,
                    TextMaskModule,
                    DropzoneModule,
                    CalendarModule
                ],
                exports: [
                    CustomSelectionComponent,
                    MaskInputComponent,
                    MutipleCheckboxComponent,
                    CustomDropzoneComponent,
                    CustomDateTimePickerComponent,
                    PhoneInputComponent,
                    CustomSelect2Component,
                    CustomCheckboxComponent,
                    UploadImagesComponent
                ],
                providers: [
                    {
                        provide: DROPZONE_CONFIG,
                        useValue: ɵ0
                    }
                ],
                declarations: [
                    CustomSelectionComponent,
                    MaskInputComponent,
                    MutipleCheckboxComponent,
                    CustomDropzoneComponent,
                    CustomDateTimePickerComponent,
                    PhoneInputComponent,
                    CustomSelect2Component,
                    CustomCheckboxComponent,
                    UploadImagesComponent
                ]
            },] }
];
export { ɵ0 };
