// angular-reactive-form/esm5/lib/directives/mutiple-checkbox/mutiple-checkbox.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, forwardRef, ElementRef, Renderer } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import * as _ from 'lodash';
var MutipleCheckboxComponent = /** @class */ (function () {
    function MutipleCheckboxComponent(element, renderer) {
        this.element = element;
        this.renderer = renderer;
        this.checkedList = [];
        this.onChange = function (val) { };
        this.onTouched = function () { };
        this.disabled = false;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    MutipleCheckboxComponent.prototype.writeValue = /**
     * @param {?} val
     * @return {?}
     */
    function (val) {
        if (val) {
            this.value = val;
        }
    };
    Object.defineProperty(MutipleCheckboxComponent.prototype, "value", {
        get: /**
         * @return {?}
         */
        function () {
            return this._value;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            this.selected = _.find(this.options, function (item) { return item.value === val; });
            this._value = val;
            this.onChange(this._value);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} fn
     * @return {?}
     */
    MutipleCheckboxComponent.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    MutipleCheckboxComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onTouched = fn;
    };
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    MutipleCheckboxComponent.prototype.setDisabledState = /**
     * @param {?} isDisabled
     * @return {?}
     */
    function (isDisabled) {
        this.disabled = isDisabled;
    };
    // tslint:disable-next-line:use-life-cycle-interface
    // tslint:disable-next-line:use-life-cycle-interface
    /**
     * @return {?}
     */
    MutipleCheckboxComponent.prototype.ngOnInit = 
    // tslint:disable-next-line:use-life-cycle-interface
    /**
     * @return {?}
     */
    function () { };
    /**
     * @param {?} option
     * @param {?} event
     * @return {?}
     */
    MutipleCheckboxComponent.prototype.onCheckboxChange = /**
     * @param {?} option
     * @param {?} event
     * @return {?}
     */
    function (option, event) {
        if (event.target.checked) {
            this.checkedList.push(option.value);
        }
        else {
            for (var i = 0; i < this.options.length; i++) {
                if (this.checkedList[i] === option.value) {
                    this.checkedList.splice(i, 1);
                }
            }
        }
        this.value = this.checkedList;
    };
    MutipleCheckboxComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'mutiple-checkbox',
                    template: "<div class=\"checkbox\" *ngFor=\"let list of options\">\n  <label>\n    <input type=\"checkbox\" value=\"{{list.value}}\" (change)=\"onCheckboxChange(list,$event)\">{{list.label}}</label>\n</div>",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(function () { return MutipleCheckboxComponent; }),
                            multi: true
                        }
                    ],
                    styles: [""]
                }] }
    ];
    MutipleCheckboxComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer }
    ]; };
    MutipleCheckboxComponent.propDecorators = {
        options: [{ type: Input }],
        _value: [{ type: Input }]
    };
    return MutipleCheckboxComponent;
}());
export { MutipleCheckboxComponent };
if (false) {
    /** @type {?} */
    MutipleCheckboxComponent.prototype.options;
    /** @type {?} */
    MutipleCheckboxComponent.prototype._value;
    /** @type {?} */
    MutipleCheckboxComponent.prototype.checkedList;
    /** @type {?} */
    MutipleCheckboxComponent.prototype.selected;
    /** @type {?} */
    MutipleCheckboxComponent.prototype.disabled;
    /** @type {?} */
    MutipleCheckboxComponent.prototype.onChange;
    /** @type {?} */
    MutipleCheckboxComponent.prototype.onTouched;
    /** @type {?} */
    MutipleCheckboxComponent.prototype.element;
    /** @type {?} */
    MutipleCheckboxComponent.prototype.renderer;
}
