// angular-reactive-form/esm2015/lib/directives/phone-input/phone-input.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { Countries } from './countries';
export class PhoneInputComponent {
    constructor() {
        this._phoneData = {
            code: '',
            value: ''
        };
        this.COUNTRIES = Countries;
        this.propagateChange = (val) => { };
    }
    /**
     * @return {?}
     */
    get phoneData() {
        return this._phoneData;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set phoneData(val) {
        this._phoneData = val;
        this.propagateChange(this._phoneData);
    }
    /**
     * @param {?} val
     * @return {?}
     */
    selectCountry(val) {
        this.phoneData = Object.assign({}, this.phoneData, { code: val });
    }
    /**
     * @param {?} val
     * @return {?}
     */
    enterPhone(val) {
        this.phoneData = Object.assign({}, this.phoneData, { value: val });
    }
    /**
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        if (value !== undefined) {
            this.phoneData = value;
        }
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.propagateChange = fn;
    }
    /**
     * @return {?}
     */
    registerOnTouched() { }
}
PhoneInputComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'app-phone-input',
                template: "<div class=\"phone-box form-group\">\n  <select #country (change)=\"selectCountry(country.value)\" class=\"phone-code\" [(ngModel)]=\"phoneData.code\">\n    <option value=\"{{ country.callingCodes[0] }}\" *ngFor=\"let country of COUNTRIES\">\n      <span>{{ country.name }}</span>\n    </option>\n  </select>\n  <div class=\"phone-number-box\">\n    <span class=\"plus-code\" *ngIf=\"phoneData.code\">(+{{ phoneData.code }}) </span>\n    <input type=\"text\" #phone (change)=\"enterPhone(phone.value)\" class=\"phone-value\" [(ngModel)]=\"phoneData.value\">\n  </div>\n</div>\n",
                providers: [{
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => PhoneInputComponent),
                        multi: true
                    }],
                styles: [".phone-box{position:relative;clear:both;border:1px solid #ccc;height:31px}.phone-box .phone-code{-webkit-appearance:none;-moz-appearance:none;appearance:none;border:0;background-color:#333;color:#fff;border-radius:0;float:left;padding:5px;margin-right:10px;width:120px}.phone-box .phone-number-box{line-height:27px}.phone-box .phone-number-box .phone-value{border:0;background-color:transparent;width:125px}"]
            }] }
];
PhoneInputComponent.propDecorators = {
    _phoneData: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    PhoneInputComponent.prototype._phoneData;
    /** @type {?} */
    PhoneInputComponent.prototype.COUNTRIES;
    /** @type {?} */
    PhoneInputComponent.prototype.propagateChange;
}
