// angular-reactive-form/esm2015/lib/directives/mutiple-checkbox/mutiple-checkbox.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, forwardRef, ElementRef, Renderer } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import * as _ from 'lodash';
export class MutipleCheckboxComponent {
    /**
     * @param {?} element
     * @param {?} renderer
     */
    constructor(element, renderer) {
        this.element = element;
        this.renderer = renderer;
        this.checkedList = [];
        this.onChange = (val) => { };
        this.onTouched = () => { };
        this.disabled = false;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    writeValue(val) {
        if (val) {
            this.value = val;
        }
    }
    /**
     * @return {?}
     */
    get value() {
        return this._value;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set value(val) {
        this.selected = _.find(this.options, item => item.value === val);
        this._value = val;
        this.onChange(this._value);
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.onChange = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    setDisabledState(isDisabled) {
        this.disabled = isDisabled;
    }
    // tslint:disable-next-line:use-life-cycle-interface
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @param {?} option
     * @param {?} event
     * @return {?}
     */
    onCheckboxChange(option, event) {
        if (event.target.checked) {
            this.checkedList.push(option.value);
        }
        else {
            for (let i = 0; i < this.options.length; i++) {
                if (this.checkedList[i] === option.value) {
                    this.checkedList.splice(i, 1);
                }
            }
        }
        this.value = this.checkedList;
    }
}
MutipleCheckboxComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'mutiple-checkbox',
                template: "<div class=\"checkbox\" *ngFor=\"let list of options\">\n  <label>\n    <input type=\"checkbox\" value=\"{{list.value}}\" (change)=\"onCheckboxChange(list,$event)\">{{list.label}}</label>\n</div>",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => MutipleCheckboxComponent),
                        multi: true
                    }
                ],
                styles: [""]
            }] }
];
MutipleCheckboxComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer }
];
MutipleCheckboxComponent.propDecorators = {
    options: [{ type: Input }],
    _value: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    MutipleCheckboxComponent.prototype.options;
    /** @type {?} */
    MutipleCheckboxComponent.prototype._value;
    /** @type {?} */
    MutipleCheckboxComponent.prototype.checkedList;
    /** @type {?} */
    MutipleCheckboxComponent.prototype.selected;
    /** @type {?} */
    MutipleCheckboxComponent.prototype.disabled;
    /** @type {?} */
    MutipleCheckboxComponent.prototype.onChange;
    /** @type {?} */
    MutipleCheckboxComponent.prototype.onTouched;
    /** @type {?} */
    MutipleCheckboxComponent.prototype.element;
    /** @type {?} */
    MutipleCheckboxComponent.prototype.renderer;
}
