// angular-reactive-form/esm2015/lib/components/reactive-form/reactive-form.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { FormControlService } from '../../services/form-control.service';
import * as _ from 'lodash';
export class ReactiveFormComponent {
    /**
     * @param {?} fcs
     */
    constructor(fcs) {
        this.fcs = fcs;
        this.groups = [];
        this.inputs = [];
        this.submitted = false;
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        /** @type {?} */
        let groups = [];
        /** @type {?} */
        let grouped = _.groupBy(this.inputs, 'group');
        // tslint:disable-next-line:forin
        for (let k in grouped) {
            groups.push({
                group: k,
                items: grouped[k],
                classes: _.head(grouped[k]).group_classes
            });
        }
        this.groups = groups;
        this.form = this.fcs.toFormGroup(this.inputs);
    }
    /**
     * @return {?}
     */
    submit() {
        this.submitted = true;
        if (this.form.invalid) {
            return false;
        }
        if (this.onSubmit !== undefined) {
            this.onSubmit.apply(null, [this.form]);
        }
    }
}
ReactiveFormComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'reactive-form',
                template: "<form (ngSubmit)=\"submit()\" [formGroup]=\"form\">\r\n  <div class=\"row\">\r\n    <div class=\"dynamic-collection\" *ngFor=\"let group of groups\" [ngClass]=\"group.classes\">\r\n      <div class=\"row\">\r\n        <div *ngFor=\"let input of group.items\" class=\"dynamic-input-group\" [ngClass]=\"input.classes\">\r\n          <dynamic-input [input]=\"input\" [form]=\"form\" [submitted]=\"submitted\"></dynamic-input>\r\n        </div>\r\n      </div>\r\n    </div>\r\n  </div>\r\n  <div class=\"dynamic-submit-btn\">\r\n    <button type=\"submit\" [ngClass]=\"{'disabled': !form.valid}\" class=\"btn btn-primary m-r-15 ripple light\">\r\n      <i class=\"icofont icofont-plus m-r-5\"></i> {{ submitText !== undefined ? submitText : 'Submit' }}\r\n    </button>\r\n  </div>\r\n</form>",
                providers: [FormControlService],
                styles: [":host .dynamic-collection{margin-bottom:10px}"]
            }] }
];
ReactiveFormComponent.ctorParameters = () => [
    { type: FormControlService }
];
ReactiveFormComponent.propDecorators = {
    inputs: [{ type: Input }],
    onSubmit: [{ type: Input }],
    submitText: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    ReactiveFormComponent.prototype.groups;
    /** @type {?} */
    ReactiveFormComponent.prototype.inputs;
    /** @type {?} */
    ReactiveFormComponent.prototype.onSubmit;
    /** @type {?} */
    ReactiveFormComponent.prototype.submitText;
    /** @type {?} */
    ReactiveFormComponent.prototype.form;
    /** @type {?} */
    ReactiveFormComponent.prototype.submitted;
    /** @type {?} */
    ReactiveFormComponent.prototype.fcs;
}
