// angular-reactive-form/esm2015/lib/directives/mask-input/mask-input.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, forwardRef, ElementRef, Renderer } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
export class MaskInputComponent {
    /**
     * @param {?} element
     * @param {?} renderer
     */
    constructor(element, renderer) {
        this.element = element;
        this.renderer = renderer;
        this.myModel = '';
        this.onChange = (val) => { };
        this.onTouched = () => { };
        this.disabled = false;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    writeValue(val) {
        if (val) {
            this.value = val;
        }
    }
    /**
     * @return {?}
     */
    get value() {
        return this._value;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set value(val) {
        this._value = val;
        this.onChange(this._value);
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.onChange = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    setDisabledState(isDisabled) {
        this.disabled = isDisabled;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        console.log(this.guide);
    }
    /**
     * @return {?}
     */
    change() {
        if (this.valueWithCharacter === true) {
            this.value = this.myModel;
        }
        else {
            this.value = this.myModel.replace(/\D+/g, '');
        }
        console.log(this.value);
    }
}
MaskInputComponent.decorators = [
    { type: Component, args: [{
                selector: 'mask-input',
                template: "<input class=\"form-control\" placeholder=\"{{placeholder}}\" [textMask]=\"{mask: inputMaskText, guide: guide}\" (change)=\"change()\" [(ngModel)]=\"myModel\" type=\"text\"/>",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => MaskInputComponent),
                        multi: true
                    }
                ],
                styles: [""]
            }] }
];
MaskInputComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer }
];
MaskInputComponent.propDecorators = {
    inputMaskText: [{ type: Input }],
    guide: [{ type: Input }],
    _value: [{ type: Input }],
    valueWithCharacter: [{ type: Input }],
    placeholder: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    MaskInputComponent.prototype.inputMaskText;
    /** @type {?} */
    MaskInputComponent.prototype.guide;
    /** @type {?} */
    MaskInputComponent.prototype._value;
    /** @type {?} */
    MaskInputComponent.prototype.valueWithCharacter;
    /** @type {?} */
    MaskInputComponent.prototype.placeholder;
    /** @type {?} */
    MaskInputComponent.prototype.myModel;
    /** @type {?} */
    MaskInputComponent.prototype.selected;
    /** @type {?} */
    MaskInputComponent.prototype.disabled;
    /** @type {?} */
    MaskInputComponent.prototype.onChange;
    /** @type {?} */
    MaskInputComponent.prototype.onTouched;
    /** @type {?} */
    MaskInputComponent.prototype.element;
    /** @type {?} */
    MaskInputComponent.prototype.renderer;
}
