// angular-reactive-form/esm5/lib/models/InputBase.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * @template T
 */
var /**
 * @template T
 */
InputBase = /** @class */ (function () {
    function InputBase(options) {
        if (options === void 0) { options = {}; }
        this.label = options.label || '';
        this.key = options.key || '';
        this.value = options.value;
        this.classes = options.classes || [];
        this.group = options.group === undefined ? 1 : options.group;
        this.group_classes = options.group_classes || ['col-12'];
        this.validators = Array.isArray(options.validators) ? options.validators : [];
        this.controlType = options.controlType || '';
    }
    return InputBase;
}());
/**
 * @template T
 */
export { InputBase };
if (false) {
    /** @type {?} */
    InputBase.prototype.label;
    /** @type {?} */
    InputBase.prototype.key;
    /** @type {?} */
    InputBase.prototype.value;
    /** @type {?} */
    InputBase.prototype.classes;
    /** @type {?} */
    InputBase.prototype.group;
    /** @type {?} */
    InputBase.prototype.group_classes;
    /** @type {?} */
    InputBase.prototype.validators;
    /** @type {?} */
    InputBase.prototype.controlType;
    /** @type {?} */
    InputBase.prototype.onChange;
    /** @type {?} */
    InputBase.prototype.accept;
    /** @type {?} */
    InputBase.prototype.allowMaxSize;
    /** @type {?} */
    InputBase.prototype.apiUpload;
    /** @type {?} */
    InputBase.prototype.uploadPath;
    /** @type {?} */
    InputBase.prototype.multiple;
    /** @type {?} */
    InputBase.prototype.options;
    /** @type {?} */
    InputBase.prototype.message;
    /** @type {?} */
    InputBase.prototype.upload_path;
    /** @type {?} */
    InputBase.prototype.acceptedFiles;
    /** @type {?} */
    InputBase.prototype.maxFilesize;
    /** @type {?} */
    InputBase.prototype.maxFiles;
    /** @type {?} */
    InputBase.prototype.showPreview;
    /** @type {?} */
    InputBase.prototype.url;
    /** @type {?} */
    InputBase.prototype.headers;
    /** @type {?} */
    InputBase.prototype.paramName;
    /** @type {?} */
    InputBase.prototype.resultTransformer;
    /** @type {?} */
    InputBase.prototype.type;
    /** @type {?} */
    InputBase.prototype.rows;
    /** @type {?} */
    InputBase.prototype.mask;
    /** @type {?} */
    InputBase.prototype.valueWithCharacter;
    /** @type {?} */
    InputBase.prototype.guide;
    /** @type {?} */
    InputBase.prototype.placeholder;
    /** @type {?} */
    InputBase.prototype.height;
    /** @type {?} */
    InputBase.prototype.plugins;
    /** @type {?} */
    InputBase.prototype.showIcon;
    /** @type {?} */
    InputBase.prototype.monthNavigator;
    /** @type {?} */
    InputBase.prototype.yearNavigator;
    /** @type {?} */
    InputBase.prototype.yearRange;
    /** @type {?} */
    InputBase.prototype.showTime;
    /** @type {?} */
    InputBase.prototype.timeOnly;
    /** @type {?} */
    InputBase.prototype.dateFormat;
    /** @type {?} */
    InputBase.prototype.tags;
}
