// angular-reactive-form/esm5/lib/directives/mask-input/mask-input.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, forwardRef, ElementRef, Renderer } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
var MaskInputComponent = /** @class */ (function () {
    function MaskInputComponent(element, renderer) {
        this.element = element;
        this.renderer = renderer;
        this.myModel = '';
        this.onChange = function (val) { };
        this.onTouched = function () { };
        this.disabled = false;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    MaskInputComponent.prototype.writeValue = /**
     * @param {?} val
     * @return {?}
     */
    function (val) {
        if (val) {
            this.value = val;
        }
    };
    Object.defineProperty(MaskInputComponent.prototype, "value", {
        get: /**
         * @return {?}
         */
        function () {
            return this._value;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            this._value = val;
            this.onChange(this._value);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} fn
     * @return {?}
     */
    MaskInputComponent.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    MaskInputComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onTouched = fn;
    };
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    MaskInputComponent.prototype.setDisabledState = /**
     * @param {?} isDisabled
     * @return {?}
     */
    function (isDisabled) {
        this.disabled = isDisabled;
    };
    /**
     * @return {?}
     */
    MaskInputComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        console.log(this.guide);
    };
    /**
     * @return {?}
     */
    MaskInputComponent.prototype.change = /**
     * @return {?}
     */
    function () {
        if (this.valueWithCharacter === true) {
            this.value = this.myModel;
        }
        else {
            this.value = this.myModel.replace(/\D+/g, '');
        }
        console.log(this.value);
    };
    MaskInputComponent.decorators = [
        { type: Component, args: [{
                    selector: 'mask-input',
                    template: "<input class=\"form-control\" placeholder=\"{{placeholder}}\" [textMask]=\"{mask: inputMaskText, guide: guide}\" (change)=\"change()\" [(ngModel)]=\"myModel\" type=\"text\"/>",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(function () { return MaskInputComponent; }),
                            multi: true
                        }
                    ],
                    styles: [""]
                }] }
    ];
    MaskInputComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer }
    ]; };
    MaskInputComponent.propDecorators = {
        inputMaskText: [{ type: Input }],
        guide: [{ type: Input }],
        _value: [{ type: Input }],
        valueWithCharacter: [{ type: Input }],
        placeholder: [{ type: Input }]
    };
    return MaskInputComponent;
}());
export { MaskInputComponent };
if (false) {
    /** @type {?} */
    MaskInputComponent.prototype.inputMaskText;
    /** @type {?} */
    MaskInputComponent.prototype.guide;
    /** @type {?} */
    MaskInputComponent.prototype._value;
    /** @type {?} */
    MaskInputComponent.prototype.valueWithCharacter;
    /** @type {?} */
    MaskInputComponent.prototype.placeholder;
    /** @type {?} */
    MaskInputComponent.prototype.myModel;
    /** @type {?} */
    MaskInputComponent.prototype.selected;
    /** @type {?} */
    MaskInputComponent.prototype.disabled;
    /** @type {?} */
    MaskInputComponent.prototype.onChange;
    /** @type {?} */
    MaskInputComponent.prototype.onTouched;
    /** @type {?} */
    MaskInputComponent.prototype.element;
    /** @type {?} */
    MaskInputComponent.prototype.renderer;
}
