// ngx-dropzone-wrapper/dist/ngx-dropzone-wrapper.es5.js
import { InjectionToken, PLATFORM_ID, NgZone, Inject, Optional, ElementRef, Renderer2, Directive, Input, Output, EventEmitter, KeyValueDiffers, Component, ViewChild, ViewEncapsulation, NgModule } from '@angular/core';
import * as Dropzone from 'dropzone';
import { isPlatformBrowser, CommonModule } from '@angular/common';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** @type {?} */
var DROPZONE_CONFIG = new InjectionToken('DROPZONE_CONFIG');
/** @type {?} */
var DropzoneEvents = [
    'error',
    'success',
    'sending',
    'canceled',
    'complete',
    'processing',
    'drop',
    'dragStart',
    'dragEnd',
    'dragEnter',
    'dragOver',
    'dragLeave',
    'thumbnail',
    'addedFile',
    'removedFile',
    'uploadProgress',
    'maxFilesReached',
    'maxFilesExceeded',
    'successMultiple',
    'sendingMultiple',
    'canceledMultiple',
    'completeMultiple',
    'processingMultiple',
    'reset',
    'queueComplete',
    'totalUploadProgress'
];
var DropzoneConfig = /** @class */ (function () {
    function DropzoneConfig(config) {
        if (config === void 0) { config = {}; }
        this.assign(config);
    }
    /**
     * @param {?=} config
     * @param {?=} target
     * @return {?}
     */
    DropzoneConfig.prototype.assign = /**
     * @param {?=} config
     * @param {?=} target
     * @return {?}
     */
    function (config, target) {
        if (config === void 0) { config = {}; }
        target = target || this;
        for (var key in config) {
            if (config[key] != null && !(Array.isArray(config[key])) &&
                typeof config[key] === 'object' && !(config[key] instanceof HTMLElement)) {
                target[key] = {};
                this.assign(config[key], target[key]);
            }
            else {
                target[key] = config[key];
            }
        }
    };
    return DropzoneConfig;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var DropzoneDirective = /** @class */ (function () {
    function DropzoneDirective(zone, renderer, elementRef, differs, platformId, defaults) {
        this.zone = zone;
        this.renderer = renderer;
        this.elementRef = elementRef;
        this.differs = differs;
        this.platformId = platformId;
        this.defaults = defaults;
        this.configDiff = null;
        this.disabled = false;
        this.DZ_ERROR = new EventEmitter();
        this.DZ_SUCCESS = new EventEmitter();
        this.DZ_SENDING = new EventEmitter();
        this.DZ_CANCELED = new EventEmitter();
        this.DZ_COMPLETE = new EventEmitter();
        this.DZ_PROCESSING = new EventEmitter();
        this.DZ_DROP = new EventEmitter();
        this.DZ_DRAGSTART = new EventEmitter();
        this.DZ_DRAGEND = new EventEmitter();
        this.DZ_DRAGENTER = new EventEmitter();
        this.DZ_DRAGOVER = new EventEmitter();
        this.DZ_DRAGLEAVE = new EventEmitter();
        this.DZ_THUMBNAIL = new EventEmitter();
        this.DZ_ADDEDFILE = new EventEmitter();
        this.DZ_REMOVEDFILE = new EventEmitter();
        this.DZ_UPLOADPROGRESS = new EventEmitter();
        this.DZ_MAXFILESREACHED = new EventEmitter();
        this.DZ_MAXFILESEXCEEDED = new EventEmitter();
        this.DZ_SUCCESSMULTIPLE = new EventEmitter();
        this.DZ_SENDINGMULTIPLE = new EventEmitter();
        this.DZ_CANCELEDMULTIPLE = new EventEmitter();
        this.DZ_COMPLETEMULTIPLE = new EventEmitter();
        this.DZ_PROCESSINGMULTIPLE = new EventEmitter();
        this.DZ_RESET = new EventEmitter();
        this.DZ_QUEUECOMPLETE = new EventEmitter();
        this.DZ_TOTALUPLOADPROGRESS = new EventEmitter();
        /** @type {?} */
        var dz = Dropzone;
        dz.autoDiscover = false;
    }
    /**
     * @return {?}
     */
    DropzoneDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (!isPlatformBrowser(this.platformId)) {
            return;
        }
        /** @type {?} */
        var params = new DropzoneConfig(this.defaults);
        params.assign(this.config); // Custom configuration
        this.renderer.addClass(this.elementRef.nativeElement, (params.maxFiles === 1) ? 'dz-single' : 'dz-multiple');
        this.renderer.removeClass(this.elementRef.nativeElement, (params.maxFiles === 1) ? 'dz-multiple' : 'dz-single');
        this.zone.runOutsideAngular(function () {
            _this.instance = new Dropzone(_this.elementRef.nativeElement, params);
        });
        if (this.disabled) {
            this.instance.disable();
        }
        // Add auto reset handling for events
        this.instance.on('success', function (result) {
            if (params.autoReset != null) {
                setTimeout(function () { return _this.reset(); }, params.autoReset);
            }
        });
        this.instance.on('error', function (error) {
            if (params.errorReset != null) {
                setTimeout(function () { return _this.reset(); }, params.errorReset);
            }
        });
        this.instance.on('canceled', function (result) {
            if (params.cancelReset != null) {
                setTimeout(function () { return _this.reset(); }, params.cancelReset);
            }
        });
        // Add native Dropzone event handling
        DropzoneEvents.forEach(function (eventName) {
            _this.instance.on(eventName.toLowerCase(), function () {
                var args = [];
                for (var _i = 0; _i < arguments.length; _i++) {
                    args[_i] = arguments[_i];
                }
                args = (args.length === 1) ? args[0] : args;
                /** @type {?} */
                var output = "DZ_" + eventName.toUpperCase();
                /** @type {?} */
                var emitter = /** @type {?} */ (_this[/** @type {?} */ (output)]);
                if (emitter.observers.length > 0) {
                    _this.zone.run(function () {
                        emitter.emit(args);
                    });
                }
            });
        });
        if (!this.configDiff) {
            this.configDiff = this.differs.find(this.config || {}).create();
            this.configDiff.diff(this.config || {});
        }
    };
    /**
     * @return {?}
     */
    DropzoneDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.instance) {
            this.zone.runOutsideAngular(function () {
                _this.instance.destroy();
            });
            this.instance = null;
        }
    };
    /**
     * @return {?}
     */
    DropzoneDirective.prototype.ngDoCheck = /**
     * @return {?}
     */
    function () {
        if (!this.disabled && this.configDiff) {
            /** @type {?} */
            var changes = this.configDiff.diff(this.config || {});
            if (changes && this.instance) {
                this.ngOnDestroy();
                this.ngOnInit();
            }
        }
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    DropzoneDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        var _this = this;
        if (this.instance && changes['disabled']) {
            if (changes['disabled'].currentValue !== changes['disabled'].previousValue) {
                if (changes['disabled'].currentValue === false) {
                    this.zone.runOutsideAngular(function () {
                        _this.instance.enable();
                    });
                }
                else if (changes['disabled'].currentValue === true) {
                    this.zone.runOutsideAngular(function () {
                        _this.instance.disable();
                    });
                }
            }
        }
    };
    /**
     * @return {?}
     */
    DropzoneDirective.prototype.dropzone = /**
     * @return {?}
     */
    function () {
        return this.instance;
    };
    /**
     * @param {?=} cancel
     * @return {?}
     */
    DropzoneDirective.prototype.reset = /**
     * @param {?=} cancel
     * @return {?}
     */
    function (cancel) {
        var _this = this;
        if (this.instance) {
            this.zone.runOutsideAngular(function () {
                _this.instance.removeAllFiles(cancel);
            });
        }
    };
    DropzoneDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[dropzone]',
                    exportAs: 'ngxDropzone'
                },] }
    ];
    /** @nocollapse */
    DropzoneDirective.ctorParameters = function () { return [
        { type: NgZone },
        { type: Renderer2 },
        { type: ElementRef },
        { type: KeyValueDiffers },
        { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] },
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [DROPZONE_CONFIG,] }] }
    ]; };
    DropzoneDirective.propDecorators = {
        disabled: [{ type: Input }],
        config: [{ type: Input, args: ['dropzone',] }],
        DZ_ERROR: [{ type: Output, args: ['error',] }],
        DZ_SUCCESS: [{ type: Output, args: ['success',] }],
        DZ_SENDING: [{ type: Output, args: ['sending',] }],
        DZ_CANCELED: [{ type: Output, args: ['canceled',] }],
        DZ_COMPLETE: [{ type: Output, args: ['complete',] }],
        DZ_PROCESSING: [{ type: Output, args: ['processing',] }],
        DZ_DROP: [{ type: Output, args: ['drop',] }],
        DZ_DRAGSTART: [{ type: Output, args: ['dragStart',] }],
        DZ_DRAGEND: [{ type: Output, args: ['dragEnd',] }],
        DZ_DRAGENTER: [{ type: Output, args: ['dragEnter',] }],
        DZ_DRAGOVER: [{ type: Output, args: ['dragOver',] }],
        DZ_DRAGLEAVE: [{ type: Output, args: ['dragLeave',] }],
        DZ_THUMBNAIL: [{ type: Output, args: ['thumbnail',] }],
        DZ_ADDEDFILE: [{ type: Output, args: ['addedFile',] }],
        DZ_REMOVEDFILE: [{ type: Output, args: ['removedFile',] }],
        DZ_UPLOADPROGRESS: [{ type: Output, args: ['uploadProgress',] }],
        DZ_MAXFILESREACHED: [{ type: Output, args: ['maxFilesReached',] }],
        DZ_MAXFILESEXCEEDED: [{ type: Output, args: ['maxFilesExceeded',] }],
        DZ_SUCCESSMULTIPLE: [{ type: Output, args: ['successMultiple',] }],
        DZ_SENDINGMULTIPLE: [{ type: Output, args: ['sendingMultiple',] }],
        DZ_CANCELEDMULTIPLE: [{ type: Output, args: ['canceledMultiple',] }],
        DZ_COMPLETEMULTIPLE: [{ type: Output, args: ['completeMultiple',] }],
        DZ_PROCESSINGMULTIPLE: [{ type: Output, args: ['processingMultiple',] }],
        DZ_RESET: [{ type: Output, args: ['reset',] }],
        DZ_QUEUECOMPLETE: [{ type: Output, args: ['queueComplete',] }],
        DZ_TOTALUPLOADPROGRESS: [{ type: Output, args: ['totalUploadProgress',] }]
    };
    return DropzoneDirective;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var DropzoneComponent = /** @class */ (function () {
    function DropzoneComponent(platformId) {
        this.platformId = platformId;
        this.disabled = false;
        this.message = 'Click or drag files to upload';
        this.placeholder = '';
        this.useDropzoneClass = true;
        this.DZ_ERROR = new EventEmitter();
        this.DZ_SUCCESS = new EventEmitter();
        this.DZ_SENDING = new EventEmitter();
        this.DZ_CANCELED = new EventEmitter();
        this.DZ_COMPLETE = new EventEmitter();
        this.DZ_PROCESSING = new EventEmitter();
        this.DZ_DROP = new EventEmitter();
        this.DZ_DRAGSTART = new EventEmitter();
        this.DZ_DRAGEND = new EventEmitter();
        this.DZ_DRAGENTER = new EventEmitter();
        this.DZ_DRAGOVER = new EventEmitter();
        this.DZ_DRAGLEAVE = new EventEmitter();
        this.DZ_THUMBNAIL = new EventEmitter();
        this.DZ_ADDEDFILE = new EventEmitter();
        this.DZ_REMOVEDFILE = new EventEmitter();
        this.DZ_UPLOADPROGRESS = new EventEmitter();
        this.DZ_MAXFILESREACHED = new EventEmitter();
        this.DZ_MAXFILESEXCEEDED = new EventEmitter();
        this.DZ_SUCCESSMULTIPLE = new EventEmitter();
        this.DZ_SENDINGMULTIPLE = new EventEmitter();
        this.DZ_CANCELEDMULTIPLE = new EventEmitter();
        this.DZ_COMPLETEMULTIPLE = new EventEmitter();
        this.DZ_PROCESSINGMULTIPLE = new EventEmitter();
        this.DZ_RESET = new EventEmitter();
        this.DZ_QUEUECOMPLETE = new EventEmitter();
        this.DZ_TOTALUPLOADPROGRESS = new EventEmitter();
    }
    /**
     * @return {?}
     */
    DropzoneComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (!isPlatformBrowser(this.platformId)) {
            return;
        }
        window.setTimeout(function () {
            DropzoneEvents.forEach(function (eventName) {
                if (_this.directiveRef) {
                    /** @type {?} */
                    var output = "DZ_" + eventName.toUpperCase();
                    /** @type {?} */
                    var directiveOutput = /** @type {?} */ (output);
                    /** @type {?} */
                    var componentOutput = /** @type {?} */ (output);
                    _this.directiveRef[directiveOutput] = /** @type {?} */ (_this[componentOutput]);
                }
            });
        }, 0);
    };
    /**
     * @return {?}
     */
    DropzoneComponent.prototype.getPlaceholder = /**
     * @return {?}
     */
    function () {
        return 'url(' + encodeURI(this.placeholder) + ')';
    };
    DropzoneComponent.decorators = [
        { type: Component, args: [{
                    selector: 'dropzone',
                    exportAs: 'ngxDropzone',
                    template: "<div class=\"dz-wrapper\" [class.dropzone]=\"useDropzoneClass\" [dropzone]=\"config\" [disabled]=\"disabled\">\n  <div class=\"dz-message\" [class.disabled]=\"disabled\" [class.dz-placeholder]=\"placeholder\">\n    <div class=\"dz-text\" [innerHTML]=\"config?.dictDefaultMessage || message\"></div>\n\n    <div *ngIf=\"placeholder\" class=\"dz-image\" [style.background-image]=\"getPlaceholder()\"></div>\n  </div>\n\n  <ng-content></ng-content>\n</div>\n",
                    encapsulation: ViewEncapsulation.None,
                    styles: ["dropzone {\n  display: block;\n  width: 100%;\n  height: auto;\n}\n\ndropzone[fxflex] {\n  display: flex;\n  flex-direction: inherit;\n  min-width: 0;\n  min-height: 0;\n\n  -webkit-box-direction: inherit;\n  -webkit-box-orient: inherit;\n}\n\ndropzone[fxflex] > .dropzone.dz-wrapper {\n  flex: 1 1 auto;\n  min-width: 0;\n  min-height: 0;\n\n  -webkit-box-flex: 1;\n}\n\ndropzone[fxlayout] {\n  align-items: inherit;\n  align-content: inherit;\n  justify-content: inherit;\n\n  -webkit-box-align: inherit;\n  -webkit-box-pack: inherit;\n}\n\ndropzone[fxlayout] > .dropzone.dz-wrapper.dz-single {\n  display: flex;\n  flex-direction: column;\n  align-items: center;\n  align-content: center;\n  justify-content: center;\n\n  -webkit-box-align: center;\n  -webkit-box-direction: column;\n  -webkit-box-orient: column;\n  -webkit-box-pack: center;\n}\n\ndropzone[fxlayout] > .dropzone.dz-wrapper.dz-multiple {\n  display: flex;\n  flex-flow: row wrap;\n  align-items: start;\n  align-content: start;\n  justify-content: space-between;\n\n  -webkit-box-align: start;\n  -webkit-box-direction: row;\n  -webkit-box-orient: row;\n  -webkit-box-pack: start;\n}\n\ndropzone > .dropzone.dz-wrapper {\n  position: relative;\n\n  overflow: auto;\n  width: 100%;\n  min-height: 0;\n  max-height: 100%;\n  padding: 0;\n  border: none;\n\n  color: #666;\n  background: transparent;\n}\n\ndropzone > .dropzone.dz-wrapper .dz-message {\n  position: relative;\n\n  display: inline-block;\n  overflow: auto;\n  width: calc(100% - 16px);\n  min-width: calc(100% - 16px);\n  min-height: 40px;\n  max-width: calc(100% - 16px);\n  max-height: 100%;\n  margin: 8px;\n  border: 2px dashed #aaa;\n\n  background-color: #eee;\n}\n\ndropzone > .dropzone.dz-wrapper .dz-message .dz-text {\n  position: absolute;\n  top: 50%;\n\n  width: 100%;\n  padding: 8px 16px;\n\n  text-align: center;\n\n  transform: translateY(-50%);\n}\n\ndropzone > .dropzone.dz-wrapper .dz-message .dz-image {\n  width: 100%;\n  height: 100%;\n\n  background-size: contain;\n  background-repeat: no-repeat;\n  background-position: 50% 50%;\n}\n\ndropzone > .dropzone.dz-wrapper .dz-message.disabled {\n  cursor: not-allowed;\n}\n\ndropzone > .dropzone.dz-wrapper .dz-message.disabled .dz-text {\n  opacity: 0.5;\n}\n\ndropzone > .dropzone.dz-wrapper .dz-message.dz-placeholder {\n  border-color: rgba(#aaa, 0);\n}\n\ndropzone > .dropzone.dz-wrapper .dz-message.dz-placeholder .dz-text {\n  position: absolute;\n  z-index: 1;\n  top: 0;\n  right: 10%;\n  left: 10%;\n\n  opacity: 0;\n\n  font-weight: bold;\n  background-color: rgba(#fff, 0.5);\n\n  transform: translateY(-50%);\n  transition:\n    filter 250ms ease-in-out,\n    opacity 250ms ease-in-out,\n    border-color 250ms ease-in-out;\n}\n\ndropzone > .dropzone.dz-wrapper .dz-message.dz-placeholder:hover:not(.disabled) {\n  border-color: #aaa;\n}\n\ndropzone > .dropzone.dz-wrapper .dz-message.dz-placeholder:hover:not(.disabled) .dz-text {\n  opacity: 1;\n}\n\ndropzone > .dropzone.dz-wrapper .dz-message.dz-placeholder:hover:not(.disabled) .dz-image {\n  -webkit-filter: blur(8px);\n\n  filter: blur(8px);\n}\n\ndropzone > .dropzone.dz-wrapper .dz-preview {\n  margin: 8px;\n}\n\ndropzone > .dropzone.dz-wrapper .dz-preview .dz-details {\n  padding: 24px;\n}\n\ndropzone > .dropzone.dz-wrapper .dz-preview .dz-progress {\n  width: 80%;\n  margin-left: -40%;\n  border: 1px solid #aaa;\n  border-radius: 4px;\n}\n\ndropzone > .dropzone.dz-wrapper .dz-preview .dz-progress .dz-upload {\n  background-color: #666;\n}\n\ndropzone > .dropzone.dz-wrapper .dz-preview .dz-filename span {\n  display: block;\n  overflow: hidden;\n  width: 100%;\n  max-width: 100%;\n\n  text-overflow: ellipsis;\n}\n\ndropzone > .dropzone.dz-wrapper .dz-preview .dz-filename span:hover {\n  overflow: visible;\n\n  white-space: normal;\n  word-wrap: break-word;\n}\n\ndropzone > .dropzone.dz-wrapper.dz-single .dz-message {\n  width: 100%;\n  height: 100%;\n}\n\ndropzone > .dropzone.dz-wrapper.dz-single.dz-started .dz-message {\n  display: none;\n}\n\ndropzone > .dropzone.dz-wrapper.dz-single .dz-preview {\n  width: calc(100% - 16px);\n  height: 100%;\n}\n\ndropzone > .dropzone.dz-wrapper.dz-single .dz-preview .dz-image {\n  width: 100%;\n  height: 100%;\n  border-radius: 0;\n}\n\ndropzone > .dropzone.dz-wrapper.dz-single .dz-preview .dz-image img {\n  display: block;\n  width: 100%;\n  height: auto;\n  margin: 0;\n}\n\ndropzone > .dropzone.dz-wrapper.dz-single .dz-error-message {\n  top: 50%;\n  left: 50%;\n\n  transform: translateX(-50%) translateY(100%);\n}\n\ndropzone > .dropzone.dz-wrapper.dz-multiple.dz-started .dz-message {\n  display: inline-block;\n}\n", "@-webkit-keyframes passing-through{0%{opacity:0;-webkit-transform:translateY(40px);-moz-transform:translateY(40px);-ms-transform:translateY(40px);-o-transform:translateY(40px);transform:translateY(40px)}30%, 70%{opacity:1;-webkit-transform:translateY(0px);-moz-transform:translateY(0px);-ms-transform:translateY(0px);-o-transform:translateY(0px);transform:translateY(0px)}100%{opacity:0;-webkit-transform:translateY(-40px);-moz-transform:translateY(-40px);-ms-transform:translateY(-40px);-o-transform:translateY(-40px);transform:translateY(-40px)}}@-moz-keyframes passing-through{0%{opacity:0;-webkit-transform:translateY(40px);-moz-transform:translateY(40px);-ms-transform:translateY(40px);-o-transform:translateY(40px);transform:translateY(40px)}30%, 70%{opacity:1;-webkit-transform:translateY(0px);-moz-transform:translateY(0px);-ms-transform:translateY(0px);-o-transform:translateY(0px);transform:translateY(0px)}100%{opacity:0;-webkit-transform:translateY(-40px);-moz-transform:translateY(-40px);-ms-transform:translateY(-40px);-o-transform:translateY(-40px);transform:translateY(-40px)}}@keyframes passing-through{0%{opacity:0;-webkit-transform:translateY(40px);-moz-transform:translateY(40px);-ms-transform:translateY(40px);-o-transform:translateY(40px);transform:translateY(40px)}30%, 70%{opacity:1;-webkit-transform:translateY(0px);-moz-transform:translateY(0px);-ms-transform:translateY(0px);-o-transform:translateY(0px);transform:translateY(0px)}100%{opacity:0;-webkit-transform:translateY(-40px);-moz-transform:translateY(-40px);-ms-transform:translateY(-40px);-o-transform:translateY(-40px);transform:translateY(-40px)}}@-webkit-keyframes slide-in{0%{opacity:0;-webkit-transform:translateY(40px);-moz-transform:translateY(40px);-ms-transform:translateY(40px);-o-transform:translateY(40px);transform:translateY(40px)}30%{opacity:1;-webkit-transform:translateY(0px);-moz-transform:translateY(0px);-ms-transform:translateY(0px);-o-transform:translateY(0px);transform:translateY(0px)}}@-moz-keyframes slide-in{0%{opacity:0;-webkit-transform:translateY(40px);-moz-transform:translateY(40px);-ms-transform:translateY(40px);-o-transform:translateY(40px);transform:translateY(40px)}30%{opacity:1;-webkit-transform:translateY(0px);-moz-transform:translateY(0px);-ms-transform:translateY(0px);-o-transform:translateY(0px);transform:translateY(0px)}}@keyframes slide-in{0%{opacity:0;-webkit-transform:translateY(40px);-moz-transform:translateY(40px);-ms-transform:translateY(40px);-o-transform:translateY(40px);transform:translateY(40px)}30%{opacity:1;-webkit-transform:translateY(0px);-moz-transform:translateY(0px);-ms-transform:translateY(0px);-o-transform:translateY(0px);transform:translateY(0px)}}@-webkit-keyframes pulse{0%{-webkit-transform:scale(1);-moz-transform:scale(1);-ms-transform:scale(1);-o-transform:scale(1);transform:scale(1)}10%{-webkit-transform:scale(1.1);-moz-transform:scale(1.1);-ms-transform:scale(1.1);-o-transform:scale(1.1);transform:scale(1.1)}20%{-webkit-transform:scale(1);-moz-transform:scale(1);-ms-transform:scale(1);-o-transform:scale(1);transform:scale(1)}}@-moz-keyframes pulse{0%{-webkit-transform:scale(1);-moz-transform:scale(1);-ms-transform:scale(1);-o-transform:scale(1);transform:scale(1)}10%{-webkit-transform:scale(1.1);-moz-transform:scale(1.1);-ms-transform:scale(1.1);-o-transform:scale(1.1);transform:scale(1.1)}20%{-webkit-transform:scale(1);-moz-transform:scale(1);-ms-transform:scale(1);-o-transform:scale(1);transform:scale(1)}}@keyframes pulse{0%{-webkit-transform:scale(1);-moz-transform:scale(1);-ms-transform:scale(1);-o-transform:scale(1);transform:scale(1)}10%{-webkit-transform:scale(1.1);-moz-transform:scale(1.1);-ms-transform:scale(1.1);-o-transform:scale(1.1);transform:scale(1.1)}20%{-webkit-transform:scale(1);-moz-transform:scale(1);-ms-transform:scale(1);-o-transform:scale(1);transform:scale(1)}}.dropzone,.dropzone *{box-sizing:border-box}.dropzone{min-height:150px;border:2px solid rgba(0,0,0,0.3);background:white;padding:20px 20px}.dropzone.dz-clickable{cursor:pointer}.dropzone.dz-clickable *{cursor:default}.dropzone.dz-clickable .dz-message,.dropzone.dz-clickable .dz-message *{cursor:pointer}.dropzone.dz-started .dz-message{display:none}.dropzone.dz-drag-hover{border-style:solid}.dropzone.dz-drag-hover .dz-message{opacity:0.5}.dropzone .dz-message{text-align:center;margin:2em 0}.dropzone .dz-preview{position:relative;display:inline-block;vertical-align:top;margin:16px;min-height:100px}.dropzone .dz-preview:hover{z-index:1000}.dropzone .dz-preview:hover .dz-details{opacity:1}.dropzone .dz-preview.dz-file-preview .dz-image{border-radius:20px;background:#999;background:linear-gradient(to bottom, #eee, #ddd)}.dropzone .dz-preview.dz-file-preview .dz-details{opacity:1}.dropzone .dz-preview.dz-image-preview{background:white}.dropzone .dz-preview.dz-image-preview .dz-details{-webkit-transition:opacity 0.2s linear;-moz-transition:opacity 0.2s linear;-ms-transition:opacity 0.2s linear;-o-transition:opacity 0.2s linear;transition:opacity 0.2s linear}.dropzone .dz-preview .dz-remove{font-size:14px;text-align:center;display:block;cursor:pointer;border:none}.dropzone .dz-preview .dz-remove:hover{text-decoration:underline}.dropzone .dz-preview:hover .dz-details{opacity:1}.dropzone .dz-preview .dz-details{z-index:20;position:absolute;top:0;left:0;opacity:0;font-size:13px;min-width:100%;max-width:100%;padding:2em 1em;text-align:center;color:rgba(0,0,0,0.9);line-height:150%}.dropzone .dz-preview .dz-details .dz-size{margin-bottom:1em;font-size:16px}.dropzone .dz-preview .dz-details .dz-filename{white-space:nowrap}.dropzone .dz-preview .dz-details .dz-filename:hover span{border:1px solid rgba(200,200,200,0.8);background-color:rgba(255,255,255,0.8)}.dropzone .dz-preview .dz-details .dz-filename:not(:hover){overflow:hidden;text-overflow:ellipsis}.dropzone .dz-preview .dz-details .dz-filename:not(:hover) span{border:1px solid transparent}.dropzone .dz-preview .dz-details .dz-filename span,.dropzone .dz-preview .dz-details .dz-size span{background-color:rgba(255,255,255,0.4);padding:0 0.4em;border-radius:3px}.dropzone .dz-preview:hover .dz-image img{-webkit-transform:scale(1.05, 1.05);-moz-transform:scale(1.05, 1.05);-ms-transform:scale(1.05, 1.05);-o-transform:scale(1.05, 1.05);transform:scale(1.05, 1.05);-webkit-filter:blur(8px);filter:blur(8px)}.dropzone .dz-preview .dz-image{border-radius:20px;overflow:hidden;width:120px;height:120px;position:relative;display:block;z-index:10}.dropzone .dz-preview .dz-image img{display:block}.dropzone .dz-preview.dz-success .dz-success-mark{-webkit-animation:passing-through 3s cubic-bezier(0.77, 0, 0.175, 1);-moz-animation:passing-through 3s cubic-bezier(0.77, 0, 0.175, 1);-ms-animation:passing-through 3s cubic-bezier(0.77, 0, 0.175, 1);-o-animation:passing-through 3s cubic-bezier(0.77, 0, 0.175, 1);animation:passing-through 3s cubic-bezier(0.77, 0, 0.175, 1)}.dropzone .dz-preview.dz-error .dz-error-mark{opacity:1;-webkit-animation:slide-in 3s cubic-bezier(0.77, 0, 0.175, 1);-moz-animation:slide-in 3s cubic-bezier(0.77, 0, 0.175, 1);-ms-animation:slide-in 3s cubic-bezier(0.77, 0, 0.175, 1);-o-animation:slide-in 3s cubic-bezier(0.77, 0, 0.175, 1);animation:slide-in 3s cubic-bezier(0.77, 0, 0.175, 1)}.dropzone .dz-preview .dz-success-mark,.dropzone .dz-preview .dz-error-mark{pointer-events:none;opacity:0;z-index:500;position:absolute;display:block;top:50%;left:50%;margin-left:-27px;margin-top:-27px}.dropzone .dz-preview .dz-success-mark svg,.dropzone .dz-preview .dz-error-mark svg{display:block;width:54px;height:54px}.dropzone .dz-preview.dz-processing .dz-progress{opacity:1;-webkit-transition:all 0.2s linear;-moz-transition:all 0.2s linear;-ms-transition:all 0.2s linear;-o-transition:all 0.2s linear;transition:all 0.2s linear}.dropzone .dz-preview.dz-complete .dz-progress{opacity:0;-webkit-transition:opacity 0.4s ease-in;-moz-transition:opacity 0.4s ease-in;-ms-transition:opacity 0.4s ease-in;-o-transition:opacity 0.4s ease-in;transition:opacity 0.4s ease-in}.dropzone .dz-preview:not(.dz-processing) .dz-progress{-webkit-animation:pulse 6s ease infinite;-moz-animation:pulse 6s ease infinite;-ms-animation:pulse 6s ease infinite;-o-animation:pulse 6s ease infinite;animation:pulse 6s ease infinite}.dropzone .dz-preview .dz-progress{opacity:1;z-index:1000;pointer-events:none;position:absolute;height:16px;left:50%;top:50%;margin-top:-8px;width:80px;margin-left:-40px;background:rgba(255,255,255,0.9);-webkit-transform:scale(1);border-radius:8px;overflow:hidden}.dropzone .dz-preview .dz-progress .dz-upload{background:#333;background:linear-gradient(to bottom, #666, #444);position:absolute;top:0;left:0;bottom:0;width:0;-webkit-transition:width 300ms ease-in-out;-moz-transition:width 300ms ease-in-out;-ms-transition:width 300ms ease-in-out;-o-transition:width 300ms ease-in-out;transition:width 300ms ease-in-out}.dropzone .dz-preview.dz-error .dz-error-message{display:block}.dropzone .dz-preview.dz-error:hover .dz-error-message{opacity:1;pointer-events:auto}.dropzone .dz-preview .dz-error-message{pointer-events:none;z-index:1000;position:absolute;display:block;display:none;opacity:0;-webkit-transition:opacity 0.3s ease;-moz-transition:opacity 0.3s ease;-ms-transition:opacity 0.3s ease;-o-transition:opacity 0.3s ease;transition:opacity 0.3s ease;border-radius:8px;font-size:13px;top:130px;left:-10px;width:140px;background:#be2626;background:linear-gradient(to bottom, #be2626, #a92222);padding:0.5em 1.2em;color:white}.dropzone .dz-preview .dz-error-message:after{content:'';position:absolute;top:-6px;left:64px;width:0;height:0;border-left:6px solid transparent;border-right:6px solid transparent;border-bottom:6px solid #be2626}\n"]
                }] }
    ];
    /** @nocollapse */
    DropzoneComponent.ctorParameters = function () { return [
        { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
    ]; };
    DropzoneComponent.propDecorators = {
        disabled: [{ type: Input }],
        config: [{ type: Input }],
        message: [{ type: Input }],
        placeholder: [{ type: Input }],
        useDropzoneClass: [{ type: Input }],
        directiveRef: [{ type: ViewChild, args: [DropzoneDirective,] }],
        DZ_ERROR: [{ type: Output, args: ['error',] }],
        DZ_SUCCESS: [{ type: Output, args: ['success',] }],
        DZ_SENDING: [{ type: Output, args: ['sending',] }],
        DZ_CANCELED: [{ type: Output, args: ['canceled',] }],
        DZ_COMPLETE: [{ type: Output, args: ['complete',] }],
        DZ_PROCESSING: [{ type: Output, args: ['processing',] }],
        DZ_DROP: [{ type: Output, args: ['drop',] }],
        DZ_DRAGSTART: [{ type: Output, args: ['dragStart',] }],
        DZ_DRAGEND: [{ type: Output, args: ['dragEnd',] }],
        DZ_DRAGENTER: [{ type: Output, args: ['dragEnter',] }],
        DZ_DRAGOVER: [{ type: Output, args: ['dragOver',] }],
        DZ_DRAGLEAVE: [{ type: Output, args: ['dragLeave',] }],
        DZ_THUMBNAIL: [{ type: Output, args: ['thumbnail',] }],
        DZ_ADDEDFILE: [{ type: Output, args: ['addedFile',] }],
        DZ_REMOVEDFILE: [{ type: Output, args: ['removedFile',] }],
        DZ_UPLOADPROGRESS: [{ type: Output, args: ['uploadProgress',] }],
        DZ_MAXFILESREACHED: [{ type: Output, args: ['maxFilesReached',] }],
        DZ_MAXFILESEXCEEDED: [{ type: Output, args: ['maxFilesExceeded',] }],
        DZ_SUCCESSMULTIPLE: [{ type: Output, args: ['successMultiple',] }],
        DZ_SENDINGMULTIPLE: [{ type: Output, args: ['sendingMultiple',] }],
        DZ_CANCELEDMULTIPLE: [{ type: Output, args: ['canceledMultiple',] }],
        DZ_COMPLETEMULTIPLE: [{ type: Output, args: ['completeMultiple',] }],
        DZ_PROCESSINGMULTIPLE: [{ type: Output, args: ['processingMultiple',] }],
        DZ_RESET: [{ type: Output, args: ['reset',] }],
        DZ_QUEUECOMPLETE: [{ type: Output, args: ['queueComplete',] }],
        DZ_TOTALUPLOADPROGRESS: [{ type: Output, args: ['totalUploadProgress',] }]
    };
    return DropzoneComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var DropzoneModule = /** @class */ (function () {
    function DropzoneModule() {
    }
    DropzoneModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule],
                    declarations: [DropzoneComponent, DropzoneDirective],
                    exports: [CommonModule, DropzoneComponent, DropzoneDirective]
                },] }
    ];
    return DropzoneModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

export { DropzoneComponent, DropzoneDirective, DROPZONE_CONFIG, DropzoneConfig, DropzoneModule };
