// ngx-dropzone-wrapper/dist/lib/dropzone.interfaces.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { InjectionToken } from '@angular/core';
/** @type {?} */
export var DROPZONE_CONFIG = new InjectionToken('DROPZONE_CONFIG');
/** @typedef {?} */
var DropzoneEvent;
export { DropzoneEvent };
/** @type {?} */
export var DropzoneEvents = [
    'error',
    'success',
    'sending',
    'canceled',
    'complete',
    'processing',
    'drop',
    'dragStart',
    'dragEnd',
    'dragEnter',
    'dragOver',
    'dragLeave',
    'thumbnail',
    'addedFile',
    'removedFile',
    'uploadProgress',
    'maxFilesReached',
    'maxFilesExceeded',
    'successMultiple',
    'sendingMultiple',
    'canceledMultiple',
    'completeMultiple',
    'processingMultiple',
    'reset',
    'queueComplete',
    'totalUploadProgress'
];
/**
 * @record
 */
export function DropzoneConfigInterface() { }
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.timeout;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.autoReset;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.errorReset;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.cancelReset;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.url;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.method;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.params;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.headers;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.init;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.accept;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.resize;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.fallback;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.renameFile;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.transformFile;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.chunksUploaded;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.withCredentials;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.previewsContainer;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.hiddenInputContainer;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.clickable;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.paramName;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.capture;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.maxFiles;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.maxFilesize;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.filesizeBase;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.acceptedFiles;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.forceFallback;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.addRemoveLinks;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.uploadMultiple;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.parallelUploads;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.resizeWidth;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.resizeHeight;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.resizeMethod;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.resizeQuality;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.resizeMimeType;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.thumbnailWidth;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.thumbnailHeight;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.thumbnailMethod;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.previewTemplate;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.autoQueue;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.autoProcessQueue;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.ignoreHiddenFiles;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.maxThumbnailFilesize;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.createImageThumbnails;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.chunking;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.chunkSize;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.retryChunks;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.forceChunking;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.retryChunksLimit;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.parallelChunkUploads;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.dictFileSizeUnits;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.dictDefaultMessage;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.dictFallbackMessage;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.dictFileTooBig;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.dictResponseError;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.dictInvalidFileType;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.dictRemoveFile;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.dictCancelUpload;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.dictUploadCanceled;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.dictFallbackText;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.dictMaxFilesExceeded;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.dictRemoveFileConfirmation;
/** @type {?|undefined} */
DropzoneConfigInterface.prototype.dictCancelUploadConfirmation;
var DropzoneConfig = /** @class */ (function () {
    function DropzoneConfig(config) {
        if (config === void 0) { config = {}; }
        this.assign(config);
    }
    /**
     * @param {?=} config
     * @param {?=} target
     * @return {?}
     */
    DropzoneConfig.prototype.assign = /**
     * @param {?=} config
     * @param {?=} target
     * @return {?}
     */
    function (config, target) {
        if (config === void 0) { config = {}; }
        target = target || this;
        for (var key in config) {
            if (config[key] != null && !(Array.isArray(config[key])) &&
                typeof config[key] === 'object' && !(config[key] instanceof HTMLElement)) {
                target[key] = {};
                this.assign(config[key], target[key]);
            }
            else {
                target[key] = config[key];
            }
        }
    };
    return DropzoneConfig;
}());
export { DropzoneConfig };
if (false) {
    /** @type {?} */
    DropzoneConfig.prototype.timeout;
    /** @type {?} */
    DropzoneConfig.prototype.autoReset;
    /** @type {?} */
    DropzoneConfig.prototype.errorReset;
    /** @type {?} */
    DropzoneConfig.prototype.cancelReset;
    /** @type {?} */
    DropzoneConfig.prototype.url;
    /** @type {?} */
    DropzoneConfig.prototype.method;
    /** @type {?} */
    DropzoneConfig.prototype.params;
    /** @type {?} */
    DropzoneConfig.prototype.headers;
    /** @type {?} */
    DropzoneConfig.prototype.init;
    /** @type {?} */
    DropzoneConfig.prototype.accept;
    /** @type {?} */
    DropzoneConfig.prototype.resize;
    /** @type {?} */
    DropzoneConfig.prototype.fallback;
    /** @type {?} */
    DropzoneConfig.prototype.renameFile;
    /** @type {?} */
    DropzoneConfig.prototype.transformFile;
    /** @type {?} */
    DropzoneConfig.prototype.chunksUploaded;
    /** @type {?} */
    DropzoneConfig.prototype.withCredentials;
    /** @type {?} */
    DropzoneConfig.prototype.previewsContainer;
    /** @type {?} */
    DropzoneConfig.prototype.hiddenInputContainer;
    /** @type {?} */
    DropzoneConfig.prototype.clickable;
    /** @type {?} */
    DropzoneConfig.prototype.paramName;
    /** @type {?} */
    DropzoneConfig.prototype.capture;
    /** @type {?} */
    DropzoneConfig.prototype.maxFiles;
    /** @type {?} */
    DropzoneConfig.prototype.maxFilesize;
    /** @type {?} */
    DropzoneConfig.prototype.filesizeBase;
    /** @type {?} */
    DropzoneConfig.prototype.acceptedFiles;
    /** @type {?} */
    DropzoneConfig.prototype.forceFallback;
    /** @type {?} */
    DropzoneConfig.prototype.addRemoveLinks;
    /** @type {?} */
    DropzoneConfig.prototype.uploadMultiple;
    /** @type {?} */
    DropzoneConfig.prototype.parallelUploads;
    /** @type {?} */
    DropzoneConfig.prototype.resizeWidth;
    /** @type {?} */
    DropzoneConfig.prototype.resizeHeight;
    /** @type {?} */
    DropzoneConfig.prototype.resizeMethod;
    /** @type {?} */
    DropzoneConfig.prototype.resizeQuality;
    /** @type {?} */
    DropzoneConfig.prototype.resizeMimeType;
    /** @type {?} */
    DropzoneConfig.prototype.thumbnailWidth;
    /** @type {?} */
    DropzoneConfig.prototype.thumbnailHeight;
    /** @type {?} */
    DropzoneConfig.prototype.thumbnailMethod;
    /** @type {?} */
    DropzoneConfig.prototype.previewTemplate;
    /** @type {?} */
    DropzoneConfig.prototype.autoQueue;
    /** @type {?} */
    DropzoneConfig.prototype.autoProcessQueue;
    /** @type {?} */
    DropzoneConfig.prototype.ignoreHiddenFiles;
    /** @type {?} */
    DropzoneConfig.prototype.maxThumbnailFilesize;
    /** @type {?} */
    DropzoneConfig.prototype.createImageThumbnails;
    /** @type {?} */
    DropzoneConfig.prototype.chunking;
    /** @type {?} */
    DropzoneConfig.prototype.chunkSize;
    /** @type {?} */
    DropzoneConfig.prototype.retryChunks;
    /** @type {?} */
    DropzoneConfig.prototype.forceChunking;
    /** @type {?} */
    DropzoneConfig.prototype.retryChunksLimit;
    /** @type {?} */
    DropzoneConfig.prototype.parallelChunkUploads;
    /** @type {?} */
    DropzoneConfig.prototype.dictFileSizeUnits;
    /** @type {?} */
    DropzoneConfig.prototype.dictDefaultMessage;
    /** @type {?} */
    DropzoneConfig.prototype.dictFallbackMessage;
    /** @type {?} */
    DropzoneConfig.prototype.dictFileTooBig;
    /** @type {?} */
    DropzoneConfig.prototype.dictResponseError;
    /** @type {?} */
    DropzoneConfig.prototype.dictInvalidFileType;
    /** @type {?} */
    DropzoneConfig.prototype.dictRemoveFile;
    /** @type {?} */
    DropzoneConfig.prototype.dictCancelUpload;
    /** @type {?} */
    DropzoneConfig.prototype.dictUploadCanceled;
    /** @type {?} */
    DropzoneConfig.prototype.dictFallbackText;
    /** @type {?} */
    DropzoneConfig.prototype.dictMaxFilesExceeded;
    /** @type {?} */
    DropzoneConfig.prototype.dictRemoveFileConfirmation;
    /** @type {?} */
    DropzoneConfig.prototype.dictCancelUploadConfirmation;
}
/** @typedef {?} */
var DropzoneUrlFunction;
export { DropzoneUrlFunction };
/** @typedef {?} */
var DropzoneMethodFunction;
export { DropzoneMethodFunction };
/** @typedef {?} */
var DropzoneParamsFunction;
export { DropzoneParamsFunction };
/** @typedef {?} */
var DropzoneHeadersFunction;
export { DropzoneHeadersFunction };
/** @typedef {?} */
var DropzoneInitFunction;
export { DropzoneInitFunction };
/** @typedef {?} */
var DropzoneFallbackFunction;
export { DropzoneFallbackFunction };
/** @typedef {?} */
var DropzoneAcceptFunction;
export { DropzoneAcceptFunction };
/** @typedef {?} */
var DropzoneResizeFunction;
export { DropzoneResizeFunction };
/** @typedef {?} */
var DropzoneRenameFileFunction;
export { DropzoneRenameFileFunction };
/** @typedef {?} */
var DropzoneTransformFileFunction;
export { DropzoneTransformFileFunction };
/** @typedef {?} */
var DropzoneChunksUploadedFunction;
export { DropzoneChunksUploadedFunction };
