// angular-reactive-form/esm2015/lib/directives/custom-select2/custom-select2.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, forwardRef, ElementRef, Renderer } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import * as _ from 'lodash';
export class CustomSelect2Component {
    /**
     * @param {?} element
     * @param {?} renderer
     */
    constructor(element, renderer) {
        this.element = element;
        this.renderer = renderer;
        this.defaultValue = [];
        this.onChange = (val) => { };
        this.onTouched = () => { };
        this.disabled = false;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    writeValue(val) {
        if (val) {
            this.value = val;
        }
    }
    /**
     * @return {?}
     */
    get value() {
        return this._value;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set value(val) {
        this._value = val;
        this.onChange(this._value);
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.onChange = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    setDisabledState(isDisabled) {
        this.disabled = isDisabled;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        // tslint:disable-next-line:prefer-const
        /** @type {?} */
        let $this = this;
        // tslint:disable-next-line:prefer-const
        /** @type {?} */
        let options = {};
        if (this.tags !== undefined && this.tags != null) {
            options.tags = this.tags;
        }
        if (this.placeholder !== undefined && this.placeholder != null) {
            options.placeholder = this.placeholder;
        }
        $('#select2-reactive-form').select2(options);
        // tslint:disable-next-line:prefer-const
        /** @type {?} */
        let oldData = $('#select2-reactive-form').select2('data');
        if (oldData != null && oldData !== undefined && !_.isEmpty(oldData)) {
            _.forEach(oldData, function (value) {
                $this.defaultValue.push(value.id);
            });
            this.value = this.defaultValue;
        }
        // console.log('Af-value', this.value);
        $('#select2-reactive-form').change(function () {
            // tslint:disable-next-line:prefer-const
            /** @type {?} */
            let value = $(this).val();
            $this.value = value;
            // console.log('Ch-value', $this.value);
        });
    }
}
CustomSelect2Component.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'app-custom-select2',
                template: "<select id=\"select2-reactive-form\" class=\"form-control\" multiple select2>\n  <option *ngFor=\"let opt of options\" value=\"{{ opt.value }}\" selected=\"{{ opt.selected }}\">{{ opt.label }}</option>\n</select>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => CustomSelect2Component),
                        multi: true
                    }
                ],
                styles: [""]
            }] }
];
CustomSelect2Component.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer }
];
CustomSelect2Component.propDecorators = {
    tags: [{ type: Input }],
    placeholder: [{ type: Input }],
    options: [{ type: Input }],
    _value: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    CustomSelect2Component.prototype.tags;
    /** @type {?} */
    CustomSelect2Component.prototype.placeholder;
    /** @type {?} */
    CustomSelect2Component.prototype.options;
    /** @type {?} */
    CustomSelect2Component.prototype._value;
    /** @type {?} */
    CustomSelect2Component.prototype.defaultValue;
    /** @type {?} */
    CustomSelect2Component.prototype.disabled;
    /** @type {?} */
    CustomSelect2Component.prototype.onChange;
    /** @type {?} */
    CustomSelect2Component.prototype.onTouched;
    /** @type {?} */
    CustomSelect2Component.prototype.element;
    /** @type {?} */
    CustomSelect2Component.prototype.renderer;
}
